### ISSUE TITLE
regression: Creating many indexes over the same column of same table causes OOM

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#11134]
### What version of Materialize are you using?

known good revision is 9611cc695
known bad revision is d115ec9

### How did you install Materialize?

Docker image

### What was the issue?

The `Indexes` scenario in the `limits` test passed previously, but is now failing with an OOM. What this test does is create 1000 indexes **over the same table**.

Previously:
- worked like a charm
- no CPU usage
- memory consumption was neglibible

Now:
- it all appears to go swimmingly until several hundred indexes are created, after which CPU and memory both spike rapidly;
- The 10+GB of memory is consumed, hence the OOM
- CPU usage spikes to 100%*num_workers and stays there for a bit before going down 
- the memory consumed is not fully reclaimed on `DROP TABLE`. Memory usage recedes, but 2Gb or thereabout remain even after `DROP` and waiting for some time

The issue:
- is reproducible even if the table is empty
- does **not** seem reproducible if you have 1000 separate tables with 1 index each

To reproduce:

Pipe the output of this script:

```
print(f"> CREATE TABLE t (f1 INTEGER)")
for i in range (1,1000):
   print(f"> CREATE INDEX i{i} ON t(f1)")
```

to `testdrive -`

### Relevant log output

Memory map post-DROP-TABLE:

![Screenshot from 2022-03-09 13-59-30](https://user-images.githubusercontent.com/1279722/157438539-43b2cd79-c0b5-44b6-bbbf-623c93ab7aaf.png)

**Comment by benesch** [comment]
Wait, there's no data in the repro??

**Comment by benesch** [comment]
@frankmcsherry or @antiguru: perhaps one of you can tell me whether this is expected? I checked out the heap profile, and it seems to be just a bunch of allocations in timely/differential. Presumably there is a fair amount of work to do to drive around 1000 table dataflows to say "nothing happened" repeatedly. 

Back when tables were materialized, I think any `CREATE INDEX` command created an index that _derived_ from the first, and only the first index had its local input driven around. And possibly that was a lot more efficient for timely/differential?

**Comment by philip-stoev** [comment]
> Wait, there's no data in the repro??

Yes, the table can be empty and the OOM still occurs.

> I checked out the heap profile,

Please note that the heap profile that was attached was taken after `DROP TABLE` plus a sufficient cool down period. So I was hoping that everything about the table , including any dataflows, would be completely wiped from memory.

**Comment by benesch** [comment]
Sorry, I meant I repro’d myself and checked out the heap profile myself with a bunch of indexes loaded up. No obviously wrong allocations, just a bunch of small allocations in timely/differential that were adding up.

**Comment by benesch** [comment]
It occurs to me: would it be wildly smarter to store only one `LocalInput` per table? So a new index would create a single-use `LocalInput`, insert all the existing table data into it, concat it with the existing table `LocalInput`, then drop the snapshot `LocalInput`. Then all the updates would flow through the one permanent `LocalInput`.

**Comment by frankmcsherry** [comment]
> Back when tables were materialized, I think any CREATE INDEX command created an index that derived from the first, and only the first index had its local input driven around. And possibly that was a lot more efficient for timely/differential?

This doesn't need to be any more efficient. When you import an arrangement, it is doing roughly the same thing as `LocalInput` means to do. It might be that `LocalInput` does it badly for reasons that aren't yet clear, but .. if you made 1k indexes in the old code you'd still have 1k things that have to get stepped forward.

**Comment by frankmcsherry** [comment]
It would be more obvious if data were involved, as one thing arrangements do is use `Rc` to wrap the backing data, allowing each import to read the same in-memory asset, rather than clone it. But, if this is happening even without data I'm not sure yet.

**Comment by frankmcsherry** [comment]
So, I think the issue is that `dataflow` is more clever than you realized. 

This command
```
print(f"> CREATE TABLE t (f1 INTEGER)")
for i in range (1,1000):
   print(f"> CREATE INDEX i{i} ON t(f1)")
```
does create 1000 indexes in the catalog. It does not necessarily create 1000 arrangements. The dataflow layer when asked to create a dataflow that imports an index, does nothing to it, and then exports an index on the same keys, is smart enough just to rebind the name of that index (well, it has been; I didn't re-confirm).

Maybe look for the regression with (note indexed expression involves `i` now).
```
print(f"> CREATE TABLE t (f1 INTEGER)")
for i in range (1,1000):
   print(f"> CREATE INDEX i{i} ON t(f1 + i)")
```
It is quite possible this won't exhibit a regression.

**Comment by frankmcsherry** [comment]
The bigger mystery to me is why, after creating the first index, is the behavior not identical to starting with an index (which afaiu is "the Previously scenario").

**Comment by frankmcsherry** [comment]
Another hypothesis is that maybe this was introduced by the fix where we now import all indexes for each dependency in `DataflowBuilder`. If you have 1k indexes, even if the same backing memory, importing them 1k times, each of 1k times, is quadratic. So .. in that case there is a bug (and I don't know what would prevent it) somewhere between `DataflowBuilder` and shipping the command to COMPUTE.

**Comment by philip-stoev** [comment]
> Maybe look for the regression with (note indexed expression involves i now):
> ...
> It is quite possible this won't exhibit a regression.

Having 1000 indexes on the same table but on 1000 different columns still reproduces the issue.

**Comment by frankmcsherry** [comment]
Sorry, I mean "won't exhibit a regression" in that the performance in previous version should also be degraded.

**Comment by philip-stoev** [comment]
No, the previous versions all seem to work fine for the case with  1000 non-identical indexes, so I have opened a separate ticket https://github.com/MaterializeInc/materialize/issues/11155 to track the second part of this regression.

**Comment by benesch** [comment]
Fixed in #11154.
